File: ios/Sources/SellwildSDK/SellwildHouseAdView.swift
// SellwildHouseAdView.swift — renders the house-ad backdrop.
//
// Two content modes, driven by `SellwildHouseAd.resolve` precedence:
//   - image:   a CMS-configured house creative (aspect-fit, full-bleed into the
//              slot; creatives are designed to the slot).
//   - listing: a Sellwild listing rendered as a compact card that mirrors the
//              feed's `ListingCardCell` — white card, photo on top, title + price
//              below. Used when no image is configured. Supplied by the feed;
//              only meaningful for the MREC slot (a 320x50 banner is too small).
//
// The view sits BEHIND the paid creative in `SellwildAdView` and is shown only
// when that creative is absent (no-fill). `SellwildAdView` hides it on paid
// fill, so a transparent or undersized creative can't let it bleed through. A
// tap opens the creative's click URL (or the listing's tap URL) via the owner.

import UIKit

final class SellwildHouseAdView: UIView {

    /// Invoked when the house ad is tapped. The owner routes this to the
    /// creative's click URL (image mode) or the listing's tap URL.
    var onTap: (() -> Void)?

    private let imageView = UIImageView()
    private let titleLabel = UILabel()
    private let priceLabel = UILabel()

    // The image URL currently being loaded. Guards against a reused view (the
    // owning SellwildAdView is pooled in feed cells) applying a stale async
    // image after the content was swapped — the feed's own cell guards the same
    // way via `currentImageURL`.
    private var pendingImageURL: String?

    // Full-bleed (image mode) vs card (listing mode) layout, toggled per content.
    private var fullBleedConstraints: [NSLayoutConstraint] = []
    private var cardConstraints: [NSLayoutConstraint] = []

    // Card-mode price color — matches ListingCardCell's default link/price blue.
    private static let priceColor = UIColor(red: 0.15, green: 0.39, blue: 0.92, alpha: 1)
    private static let titleColor = UIColor(red: 0.07, green: 0.09, blue: 0.15, alpha: 1)

    override init(frame: CGRect) {
        super.init(frame: frame)
        clipsToBounds = true
        layer.cornerRadius = 12

        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.clipsToBounds = true
        addSubview(imageView)

        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.font = .systemFont(ofSize: 15, weight: .bold)
        titleLabel.textColor = Self.titleColor
        titleLabel.numberOfLines = 2
        titleLabel.isHidden = true
        addSubview(titleLabel)

        priceLabel.translatesAutoresizingMaskIntoConstraints = false
        priceLabel.font = .systemFont(ofSize: 18, weight: .bold)
        priceLabel.textColor = Self.priceColor
        priceLabel.isHidden = true
        addSubview(priceLabel)

        // Common: image pinned to the top edges in both modes.
        NSLayoutConstraint.activate([
            imageView.topAnchor.constraint(equalTo: topAnchor),
            imageView.leadingAnchor.constraint(equalTo: leadingAnchor),
            imageView.trailingAnchor.constraint(equalTo: trailingAnchor),
        ])

        // Image mode: photo fills the whole slot.
        fullBleedConstraints = [
            imageView.bottomAnchor.constraint(equalTo: bottomAnchor),
        ]

        // Card mode: photo occupies the top ~60% of the slot; title + price sit
        // on the white card below it (mirrors ListingCardCell's stacking).
        cardConstraints = [
            imageView.heightAnchor.constraint(equalTo: heightAnchor, multiplier: 0.6),
            titleLabel.topAnchor.constraint(equalTo: imageView.bottomAnchor, constant: 8),
            titleLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 12),
            titleLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -12),
            priceLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 4),
            priceLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 12),
            priceLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -12),
        ]

        addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(tapped)))
        isUserInteractionEnabled = true
    }

    required init?(coder: NSCoder) { fatalError("Use init(frame:)") }

    @objc private func tapped() { onTap?() }

    /// Render a CMS house image (aspect-fit; creatives are designed to the slot).
    func showImage(_ creative: SellwildHouseAdCreative) {
        setCardLayout(false)
        backgroundColor = .clear
        imageView.contentMode = .scaleAspectFit
        imageView.backgroundColor = .clear
        loadImage(creative.imageURL)
    }

    /// Render a Sellwild listing as a compact card (photo on top, title + price
    /// below on a white card) so it reads like the feed's listing cards.
    func showListing(_ listing: SellwildListing, config: SellwildConfig) {
        setCardLayout(true)
        backgroundColor = .white
        imageView.contentMode = .scaleAspectFill
        imageView.backgroundColor = UIColor(white: 0.93, alpha: 1)
        titleLabel.text = listing.title
        priceLabel.text = Self.formatPrice(currency: listing.currency, price: listing.price)
        loadImage(listing.photos?.first?.url)
    }

    /// Switch between full-bleed (image) and card (listing) layout, showing the
    /// title/price chrome only in card mode.
    private func setCardLayout(_ card: Bool) {
        titleLabel.isHidden = !card
        priceLabel.isHidden = !card
        NSLayoutConstraint.deactivate(card ? fullBleedConstraints : cardConstraints)
        NSLayoutConstraint.activate(card ? cardConstraints : fullBleedConstraints)
    }

    /// Load `urlString` into the image view, ignoring a result that arrives after
    /// the content was swapped (reused view). Clears any prior image first so a
    /// stale photo never lingers under new content.
    private func loadImage(_ urlString: String?) {
        pendingImageURL = urlString
        imageView.image = nil
        guard let urlString, !urlString.isEmpty else { return }
        SellwildHouseAd.loadImage(urlString) { [weak self] image in
            guard let self, self.pendingImageURL == urlString else { return }
            self.imageView.image = image
        }
    }

    private static func formatPrice(currency: String?, price: String?) -> String {
        guard let p = price, let value = Double(p) else { return "" }
        let sym: String
        switch currency?.uppercased() {
        case "EUR": sym = "€"
        case "GBP": sym = "£"
        default:    sym = "$"
        }
        return value.truncatingRemainder(dividingBy: 1) == 0
            ? "\(sym)\(Int(value))"
            : String(format: "%@%.2f", sym, value)
    }
}

File: android/src/main/kotlin/com/sellwild/sdk/SellwildHouseAdView.kt
// SellwildHouseAdView.kt — renders the house-ad backdrop.
//
// Two content modes, driven by SellwildHouseAd.resolve precedence:
//   - image:   a CMS-configured house creative (fit, full-bleed into the slot;
//              creatives are designed to the slot).
//   - listing: a Sellwild listing rendered as a compact card that mirrors the
//              feed's real listing card — white card, photo on top, title + price
//              below. Used when no image is configured. Supplied by the feed;
//              only meaningful for the MREC slot (a 320x50 banner is too small).
//
// The view sits BEHIND the paid creative in SellwildAdView and is shown only
// when that creative is absent (no-fill). SellwildAdView hides it on paid fill,
// so a transparent or undersized creative can't let it bleed through. A tap
// opens the creative's click URL (or the listing's tap URL) via the owner.

package com.sellwild.sdk

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

internal class SellwildHouseAdView(context: Context) : FrameLayout(context) {

    /** Invoked when the house ad is tapped; the owner routes it to the URL. */
    var onTap: (() -> Unit)? = null

    private val imageView: ImageView
    private val textArea: LinearLayout
    private val titleView: TextView
    private val priceView: TextView
    private val card: LinearLayout

    // The image URL currently being loaded. Guards against a reused view (the
    // owning SellwildAdView is pooled in feed rows) applying a stale async image
    // after the content was swapped — the feed's own cell guards the same way.
    private var pendingImageUrl: String? = null

    init {
        layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        )
        isClickable = true
        isFocusable = true
        setOnClickListener { onTap?.invoke() }

        imageView = ImageView(context)

        titleView = TextView(context).apply {
            textSize = 15f
            maxLines = 2
            setTypeface(typeface, Typeface.BOLD)
            setTextColor(Color.rgb(18, 23, 38))          // ListingCardCell title
        }
        priceView = TextView(context).apply {
            textSize = 18f
            setTypeface(typeface, Typeface.BOLD)
            setTextColor(Color.rgb(38, 99, 235))         // ListingCardCell price/link blue
        }
        val padH = dp(12)
        textArea = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(padH, dp(8), padH, dp(12))
            addView(titleView)
            addView(priceView)
            visibility = GONE
        }

        // Vertical card: photo on top (fills the space above), text below. In
        // image mode the text area is GONE so the photo fills the whole slot.
        card = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
            addView(
                imageView,
                LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1f),
            )
            addView(
                textArea,
                LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                ),
            )
        }
        addView(card)
    }

    /** Render a CMS house image (fit into the slot; creatives are slot-sized). */
    fun showImage(creative: SellwildHouseAd.Creative) {
        textArea.visibility = GONE
        card.setBackgroundColor(Color.TRANSPARENT)
        imageView.scaleType = ImageView.ScaleType.FIT_CENTER
        imageView.setBackgroundColor(Color.TRANSPARENT)
        loadImage(creative.imageUrl)
    }

    /** Render a Sellwild listing as a compact card (photo on top, title + price
     *  below on a white card) so it reads like the feed's listing cards. */
    fun showListing(listing: SellwildListing, config: SellwildConfig) {
        textArea.visibility = VISIBLE
        card.setBackgroundColor(Color.WHITE)
        imageView.scaleType = ImageView.ScaleType.CENTER_CROP
        imageView.setBackgroundColor(Color.parseColor("#EEEEEE"))
        titleView.text = listing.title
        priceView.text = formatPrice(listing.currency, listing.price)
        loadImage(listing.primaryPhotoUrl)
    }

    /** Load [url] into the image view, ignoring a result that arrives after the
     *  content was swapped (reused view). Clears any prior image first. */
    private fun loadImage(url: String?) {
        pendingImageUrl = url
        imageView.setImageDrawable(null)
        if (url.isNullOrEmpty()) return
        SellwildHouseAd.loadImage(context, url) { bmp ->
            if (pendingImageUrl == url) bmp?.let { imageView.setImageBitmap(it) }
        }
    }

    private fun dp(value: Int): Int =
        (value * context.resources.displayMetrics.density).toInt()

    private fun formatPrice(currency: String?, price: String?): String {
        val value = price?.toDoubleOrNull() ?: return ""
        val sym = when (currency?.uppercase()) {
            "EUR" -> "€"
            "GBP" -> "£"
            else -> "$"
        }
        return if (value % 1.0 == 0.0) "$sym${value.toInt()}" else "$sym${"%.2f".format(value)}"
    }
}
